---
title: "ClusterFoldSimilarity: comparing different single-cell experiments"
author: 
- name: Óscar González-Velasco
  affiliation: Division of Applied Bioinformatics, German Cancer Research Center DKFZ 
package: SingleCellExperiment
output:
  BiocStyle::html_document:
    toc_float: true
vignette: >
  %\VignetteIndexEntry{ClusterFoldSimilarity: }
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r options, include=FALSE, echo=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(eval=TRUE,warning=FALSE, error=FALSE, message=FALSE)
```

# Introduction

Comparing single-cell data across different datasets, samples and batches has demonstrated to be challenging. `ClusterFoldSimilarity` aims to solve the complexity of comparing different single-cell data by calculating similarity scores between clusters (or user-defined groups) from two or more single-cell experiments. It does it by finding similar fold-change patterns across cell groups with a common set of features (e.g. genes). It also reports the top genes that contributed the most to the similarity value, acting as a feature selection tool.

`ClusterFoldSimilarity` can be used with single-cell RNA-Seq data, single-cell ATAC-Seq data, or more broadly, with continuous numerical data that shows signal changes across a set of common features from different groups. It is compatible with the most used single-cell objects: `r Biocpkg("Seurat")` and `r Biocpkg("SingleCellExperiment")`.

The output is a table that contains the similarity values for each pair of clusters from all datasets. `ClusterFoldSimilarity` also includes various plotting functions to help visualize the similarity scores.

# Using ClusterFoldSimilarity to find similar clusters across datasets

Typically `ClusterFoldSimilarity` will receive as input either a list of two or more `r Biocpkg("Seurat")` objects or a list of two or more `r Biocpkg("SingleCellExperiment")` objects, containing already processed data: e.g. filtered, normalized and clustered. (*PLEASE NOTE: this package is intended to be used with high-end-analyzed data, the better the pre-analysis the better the results `ClusterFoldSimilarity` will obtain, this includes: normalizing and taking care of local technical noise effects, removing non-variant data or selecting the top variant features, removing 0 expression features, reasonable number of clusters, etc.*)

`ClusterFoldSimilarity` will automatically look inside these objects for **normalized data** ( `GetAssayData(assay, slot = "data")` for `r Biocpkg("Seurat")` or `normcounts()` for `r Biocpkg("SingleCellExperiment")` ) and **cluster or label information** ( `Idents()` for `r Biocpkg("Seurat")` and `colLabels()` for `r Biocpkg("SingleCellExperiment")` ).


For the purpose of this example, we will use the package scRNAseq that contains several single-cell datasets, including samples from mouse brain.

```{r construct }
library(Seurat)
library(scRNAseq)

# Mouse brain single-cell RNA-seq 1 from Romanov et. al.
mouse.brain.romanov <- scRNAseq::RomanovBrainData(ensembl = TRUE)
colnames(mouse.brain.romanov) <- colData(mouse.brain.romanov)$cellID
rownames(colData(mouse.brain.romanov)) <- colData(mouse.brain.romanov)$cellID
singlecell.1.seurat <- CreateSeuratObject(counts = counts(mouse.brain.romanov),meta.data = as.data.frame(colData(mouse.brain.romanov)))

# Mouse brain single-cell RNA-seq 2 from Zeisel et. al.
mouse.brain.zei <- scRNAseq::ZeiselBrainData(ensembl = TRUE)
singlecell.2.seurat <- CreateSeuratObject(counts = counts(mouse.brain.zei),meta.data = as.data.frame(colData(mouse.brain.zei)))
```

Normalize and identify variable features for each dataset independently
*Note: these steps should be done tailored to each independent dataset, here we apply the same parameters for the sake of simplicity:*

```{r}
# Create a list with the unprocessed single-cell datasets
singlecell.object.list <- list(singlecell.1.seurat,singlecell.2.seurat)
# Apply the same processing to each dataset and return a list of single-cell analysis
singlecell.object.list <- lapply(X = singlecell.object.list, FUN = function(x){
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 1000)
  x <- ScaleData(x,features = VariableFeatures(x))
  x <- RunPCA(x, features = VariableFeatures(object = x))
  x <- FindNeighbors(x, dims = seq(10))
  x <- FindClusters(x, resolution = 0.1)
})
```

Once we have all of our single-cell datasets analyzed independently, we can compute the similarity values. `cluster_fold_similarity()` takes as arguments:

  - `sce_list`: a list of single-cell objects (mandatory) either of class Seurat or of class SingleCellExperiment.
  - `top_n`: the top n most similar clusters to report from each pair of clusters (default: `1`, the top most similar cluster). If set to `Inf` it will return all the values from all the cluster-pairs.
  - `top_n_genes`: the top n genes that contributes to the similarity between the pair of clusters (default: `1`, the top contributing gene).

```{r}
library(ClusterFoldSimilarity)
similarity.table <- cluster_fold_similarity(sce_list = singlecell.object.list,top_n = 1)
head(similarity.table)
```

By default, `cluster_fold_similarity()` will plot a graph network that visualizes the connections between the clusters from the different datasets using the similarity table that has been obtained, the arrows point in the direction of the similarity (dataset_l:cluster_l -> dataset_r:cluster_r); it can be useful for identifying relationships between groups of clusters and cell-populations that tend to be more similar.

In this example, as we have information regarding cell-type labels, we can check how the cell types match by calculating the most abundant cell type on each of the similar clusters:

```{r}
label1 <- "level1.class" # name of label of data 1
label2 <- "level1class" # name of label of data 2

apply(similarity.table[similarity.table$dataset_l == 1,],1,function(x){
  n1 <- names(which.max(table(singlecell.object.list[[as.numeric(x["dataset_l"])]]@meta.data[singlecell.object.list[[as.numeric(x["dataset_l"])]]@meta.data$seurat_clusters == x["cluster_l"],label1])))
  n2 <- names(which.max(table(singlecell.object.list[[as.numeric(x["dataset_r"])]]@meta.data[singlecell.object.list[[as.numeric(x["dataset_r"])]]@meta.data$seurat_clusters == x["cluster_r"],label2])))
  return(paste("dataset 1 cluster",x["cluster_l"],"top cell.type:",n1,"VS dataset 2 cluster",x["cluster_r"],"top cell.type:",n2))
  })
```

If we suspect that clusters could be related with more than one cluster of other datasets, we can retrieve the top n similarities for each cluster: 

```{r}
# Retrieve the top 3 similar cluster for each of the clusters:
similarity.table.3top <- cluster_fold_similarity(sce_list = singlecell.object.list,top_n = 3)
head(similarity.table.3top)
```

If we are interested on the features that contribute the most to the similarity, we can retrieve the top n features: 

```{r}
# Retrieve the top 5 features that contribute the most to the similarity between each pair of clusters:
similarity.table.5top.features <- cluster_fold_similarity(sce_list = singlecell.object.list,top_n_genes = 5)
head(similarity.table.5top.features, n=10)
```

# Retrieving all the similarity values

Sometimes it is useful to retrieve all the similarity values for downstream analysis (e.g. identify more than one cluster that is similar to a cluster of interest, finding the most dissimilar clusters, etc). To obtain all the values, we need to specify `top_n=Inf`:

```{r}
similarity.table.all.values <- cluster_fold_similarity(sce_list = singlecell.object.list,top_n = Inf)
dim(similarity.table.all.values)
```

It can be convenient to create a matrix with all the similarity values from the comparison of two datasets:

```{r}
library(dplyr)
dataset1 <- 1
dataset2 <- 2
similarity.table.2 <- similarity.table.all.values %>% 
                      filter(dataset_l == dataset1 & dataset_r == dataset2) %>% 
                      arrange(desc(as.numeric(cluster_l)),as.numeric(cluster_r))
cls <- unique(similarity.table.2$cluster_l)
cls2 <- unique(similarity.table.2$cluster_r)
similarity.matrix.all <- t(matrix(similarity.table.2$similarity_value,ncol=length(unique(similarity.table.2$cluster_l))))
rownames(similarity.matrix.all) <- cls
colnames(similarity.matrix.all) <- cls2
similarity.matrix.all
```

# Using ClusterFoldSimilarity for cell annotation:

If we want to use a specific single-cell dataset for annotation (from which we know a ground-truth label, e.g. cell type, cell cycle, treatment... etc.), we can use that label to directly compare the single-cell datasets.

For this example we will use the dataset from Romanov et. al. that contains cell-type annotations from mouse brain samples:

```{r}
# The name of the label we want to use for annotation:
cell.label.name <- "level1.class"
# Visualize the label we are using as ground-truth:
table(singlecell.object.list[[1]]@meta.data[,cell.label.name])
# Set the group label in the Seurat object:
Idents(singlecell.object.list[[1]]) <- cell.label.name

similarity.table.cell.labeling <- cluster_fold_similarity(sce_list = singlecell.object.list,
                                                          sample_names = c("labeled cells","new samples"),
                                                          top_n = 1)
```

We can visualize all the similarities for each cluster to the annotated cell-groups:

```{r}
similarity.table.cell.labeling.all <- cluster_fold_similarity(sce_list = singlecell.object.list,top_n = Inf,
                                                              sample_names = c("labeled cells","new samples"))
# We can select which dataset to plot in the Y-axis:
ClusterFoldSimilarity::similarity_heatmap(similarity_table = similarity.table.cell.labeling.all,main_dataset = "new samples")
```

# Working with multiple datasets:

With `ClusterFoldSimilarity` we can compare any number of independent single-cell datasets:

```{r}
# Example with 3 datasets: we split the single-cell RNA-seq from Zeisel et. al. by tissue:
singlecell.object.list.split <- Seurat::SplitObject(singlecell.2.seurat,split.by = "tissue")
singlecell.object.list.split <- lapply(X = singlecell.object.list.split, FUN = function(x){
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 1000)
  x <- ScaleData(x,features = VariableFeatures(x))
  x <- RunPCA(x, features = VariableFeatures(object = x))
  x <- FindNeighbors(x, dims = 1:10)
  x <- FindClusters(x, resolution = 0.1)
})

singlecell.object.list.3.datasets <- c(singlecell.object.list[[1]],singlecell.object.list.split)
similarity.table.3.datasets <- cluster_fold_similarity(sce_list = singlecell.object.list.3.datasets,sample_names = c("romanov","zei.cortex","zei.hippo"))
```

We can plot the heatmap of all the similarity values for any number of independent datasets:

```{r}
similarity.table.3.datasets <- cluster_fold_similarity(sce_list = singlecell.object.list.3.datasets,
                                                       sample_names = c("romanov","zei.cortex","zei.hippo"),
                                                       top_n = Inf)
```

# Similarity score calculation

`ClusterFoldSimilarity` does not need to merge or harmonize the data across the datasets that we aim to analyze, which makes it less prone to data-loss or noise, and that is typical of some of these methods. The similarity value is based on the fold-changes between clusters/groups of cells defined by the user. These fold-changes from different independent datasets are combined using a pairwise dot product approach, after adding these products and applying a fold-change concordance weight, a similarity value is calculate for each of the clusters of each of the datasets present.

```{r setup}
library(ClusterFoldSimilarity)
```

# Session information {-}

```{r}
sessionInfo()
```